kong/concurrency.lua

```diff
@@ -109,6 +109,15 @@ function concurrency.with_coroutine_mutex(opts, fn)
     semaphore:post(1)
   end
 
+  -- to resolve deadlock issue in case the worker event thread
+  -- associated with `pcall(fn)` below got killed
+  -- and the `:post(1)` followed is skipped.
+  if semaphore:count() <= 0 then
```

Comment:
I think I need some help understanding the `<=`, [the docs](https://github.com/openresty/lua-resty-core/blob/master/lib/ngx/semaphore.md#count) say:

> When the returned number is negative, it means the number of "light threads" waiting on this semaphore.

If we are trying to detect the scenario where there are *no* light threads associated with the semaphore, should it just be `== 0`?

Reply:
@samugi good point.

The `if` is detecting "no semaphores were posted", and then zero or more light threads would be blocked.

1. If the value is zero, then no light threads are waiting nor posting. In this case, we should post 1 to prepare for next wait.
2. If the value is less than zero, then there are threads already blocked. We should post at least 1. I am considering if we should post the number of waiting threads, like this:

    ```lua
    local waiting_count = semaphore:count()
    
    if waiting_count <=0 then
      semaphore:post(-waiting_count)
    end
    ```

Reply:
if I understood correctly, `1` is the use case that this PR would be fixing, in case of `2`, then the threads were not killed and we would expect the normal workflow to release the lock after the `pcall` has returned, please correct me if I'm wrong.

Also, apparently `post` [only accepts positive numbers](https://github.com/openresty/lua-resty-core/blob/f9add683874e4241d0d93446022af89adf471299/lib/ngx/semaphore.lua#L174).

Reply:
In case of `2`, it means that those threads are already locked at here `local lok, err = semaphore:wait(timeout)`, we should post 1.

For example, a thread is locked at `local lok, err = semaphore:wait(timeout)`, the `:count()` will return `-1`.

Reply:
I was wondering if there is a risk that by doing this we unlock threads that are supposed to be waiting, defeating the purpose of the mutex